### PR TITLE
zkvm improvements

### PR DESCRIPTION
**Pull request by prajwolrg**
## Description

* Introduced a new `ProofWithMetadata` that includes unique Id and remote Id
* Added the capability to run simulations within the ZKVM environment, allowing for better testing and validation of proof generation.
* Updated SP1-related tests to support the newly added simulation feature, enhancing the test coverage and reliability.
* Introduced logging functionality in SP1 ZKVM tests to provide better visibility into the test execution and debugging process.
* Moved the calculation of transaction filter rules from BTC blockspace logic to the L1 batch to improve proving cycles.

### Type of Change

-   [x] Bug fix (non-breaking change which fixes an issue)
-   [x] New feature/Enhancement (non-breaking change which adds functionality or enhances an existing one)
-   [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
-   [ ] Documentation update
-   [x] Refactor

## Checklist

-   [x] I have performed a self-review of my code.
-   [x] I have commented my code where necessary.
-   [x] I have updated the documentation if needed.
-   [x] My changes do not introduce new warnings.
-   [x] I have added tests that prove my changes are effective or that my feature works.
-   [x] New and existing tests pass with my changes.

### CODE REVIEW

File: crates/proof-impl/btc-blockspace/src/logic.rs [hunk 1]
```diff
@@ -2,59 +2,64 @@
...
-    } = input;
-    assert!(check_merkle_root(block));
+pub fn process_blockspace_proof(
+    serialized_block: &[u8],
+    cred_rule: &CredRule,
+    serialized_tx_filters: &[u8],
+) -> BlockspaceProofOutput {
+    let block: Block = bitcoin::consensus::deserialize(serialized_block).unwrap();
+    assert!(check_merkle_root(&block));
     // assert!(check_witness_commitment(block));
```
Comment (MdTeach) [inline crates/proof-impl/btc-blockspace/src/logic.rs:27]: For `assert!(check_witness_commitment(block));` can this be removed or its something we have disabled for now. In that case we can leave a comment there.

File: crates/zkvm/adapters/risc0/src/prover.rs [hunk 2]
```diff
@@ -60,13 +60,31 @@ impl ZKVMHost for RiscZeroHost {
         // let proof_info = prover.prove_session(&ctx, &session)?;
         let proof_info = prover.prove_with_opts(prover_input, &self.elf, &opts)?;
 
+        let proof_id = "todo".to_string();
```
Comment (MdTeach) [inline crates/zkvm/adapters/risc0/src/prover.rs:63]: Although we are not using r0, we can have better name here with todo comment.

File: crates/zkvm/adapters/sp1/src/verifier.rs [hunk 3]
```diff
@@ -117,8 +117,13 @@ mod tests {
             .try_as_groth_16()
             .expect("Failed to convert proof to Groth16")
             .raw_proof;
-        let groth16_proof =
-            Proof::new(hex::decode(&groth16_proof_bytes).expect("Failed to decode Groth16 proof"));
+
+        let groth16_proof_id = "todo".to_string();
```
Comment (MdTeach) [inline crates/zkvm/adapters/sp1/src/verifier.rs:121]: Here, we can have better naming.

File: functional-tests/fn_client_restart.py [hunk 4]
```diff
@@ -51,7 +51,9 @@ def main(self, ctx: flexitest.RunContext):
 def check_already_sent_proof(seqrpc):
     try:
         # Proof for checkpoint 0 is already sent
-        seqrpc.strataadmin_submitCheckpointProof(0, "abc123")
+        seqrpc.strataadmin_submitCheckpointProof(
+            0, "000000000000000000"
```
Comment (MdTeach) [inline functional-tests/fn_client_restart.py:55]: So, `000000000000000000` is the default proof now ?
Earlier it was empty string.

File: provers/risc0/tests/checkpoint.rs [hunk 5]
```diff
@@ -1,10 +1,10 @@
-#[cfg(feature = "prover")]
+// #[cfg(feature = "prover")]
```
Comment (MdTeach) [inline provers/risc0/tests/checkpoint.rs:1]: We don't need feature prover here now ?

File: provers/risc0/tests/l1-batch.rs [hunk 6]
```diff
@@ -28,7 +27,7 @@ mod test {
         };
         let prover = RiscZeroHost::init(GUEST_RISC0_BTC_BLOCKSPACE_ELF.into(), prover_options);
 
-        let btc_chain = get_btc_chain();
+        // let btc_chain = get_btc_chain();
```
Comment (MdTeach) [inline provers/risc0/tests/l1-batch.rs:30]: We can remove this unwanted comment.

File: functional-tests/fn_client_restart.py (hunk 4 above)
Comment (prajwolrg) [reply functional-tests/fn_client_restart.py:55]: Yes. Earlier `Proof` only included the raw proof. Now it also includes the id.

File: crates/proof-impl/btc-blockspace/src/logic.rs (hunk 1 above)
Comment (storopoli) [reply crates/proof-impl/btc-blockspace/src/logic.rs:27]: leave a FIXME TODO etc.

File: crates/state/src/batch.rs [hunk 7]
```diff
@@ -70,9 +70,9 @@ impl SignedBatchCheckpoint {
         &self.inner
     }
 
-    pub fn verify_sig(&self, pub_key: Buf32) -> bool {
+    pub fn verify_sig(&self, pub_key: &Buf32) -> bool {
         let msg = self.checkpoint().get_sighash();
-        verify_schnorr_sig(&self.signature, &msg, &pub_key)
+        verify_schnorr_sig(&self.signature, &msg, pub_key)
```
Comment (storopoli) [inline crates/state/src/batch.rs:75]: Why do we need to take ownership of `pub_key` can't we verify it by reference?
A lot of copies here... might have some perf issues in the future

Comment (prajwolrg) [reply crates/state/src/batch.rs:75]: Looks good now?

Comment (storopoli) [reply crates/state/src/batch.rs:75]: Yeah

**Review by MdTeach — APPROVED** [review]

File: provers/sp1/tests/helpers/btc.rs [hunk 8]
```diff
@@ -19,22 +22,30 @@ impl BtcBlockProofGenerator {
...
         let serialized_block = serialize(block);
 
-        let input = SP1ProofInputBuilder::new()
-            .write(rollup_params)?
+        info!("Building input for BTC Blockspace Proof");
+        SP1ProofInputBuilder::new()
+            .write(&rollup_params.cred_rule)?
             .write_serialized(&serialized_block)?
-            .build()?;
+            .write_serialized(&serialized_tx_filters)?
```
Comment (MdTeach) [inline provers/sp1/tests/helpers/btc.rs:37]: Need to add these changes to the prover manager

File: provers/sp1/tests/helpers/l1_batch.rs [hunk 9]
```diff
@@ -26,31 +27,49 @@ impl L1BatchProofGenerator {
...
 
         for height in start_height..=end_height {
             let block = btc_chain.get_block(height);
-            let (proof, vk) = self.btc_proof_generator.get_proof(block, prover_options)?;
+            let (proof, vk) = self.btc_proof_generator.get_proof(block, &prover_options)?;
+            debug!(%height, "Fetched BTC Block proof for agg input");
             input_builder.write_proof(AggregationInput::new(proof, vk))?;
         }
+        input_builder.build()
+    }
```
Comment (MdTeach) [inline provers/sp1/tests/helpers/l1_batch.rs:62]: Need to make similar changes on the prover manager input builder of L1 batch

**Review by MdTeach — CHANGES_REQUESTED** [review]

**Comment by prajwolrg** [comment]
Closing in favor of #390

**Comment by prajwolrg** [comment]
Closing in favor of #390

**Pull request closed without merging**